libcaf_net/caf/net/octet_stream/transport.hpp

```diff
@@ -30,194 +27,32 @@ class CAF_NET_EXPORT transport : public socket_event_layer,
-
-  // -- implementation of octet_stream::lower_layer ----------------------------
-
-  socket_manager* manager() noexcept override;
-
-  bool can_send_more() const noexcept override;
-
-  void configure_read(receive_policy policy) override;
-
-  void begin_output() override;
-
-  byte_buffer& output_buffer() override;
-
-  bool end_output() override;
-
-  bool is_reading() const noexcept override;
+  static std::unique_ptr<transport> make(std::unique_ptr<policy> policy,
+                                         upper_layer_ptr up);
 
-  void write_later() override;
-
-  void shutdown() override;
-
-  void switch_protocol(upper_layer_ptr) override;
-
-  bool switching_protocol() const noexcept override;
+  static std::unique_ptr<transport> make(stream_socket fd, upper_layer_ptr up);
 
   // -- properties -------------------------------------------------------------
 
-  auto& read_buffer() noexcept {
-    return read_buf_;
-  }
-
-  const auto& read_buffer() const noexcept {
-    return read_buf_;
-  }
-
-  auto& write_buffer() noexcept {
-    return write_buf_;
-  }
-
-  const auto& write_buffer() const noexcept {
-    return write_buf_;
-  }
-
-  auto& upper_layer() noexcept {
-    return *up_;
-  }
-
-  const auto& upper_layer() const noexcept {
-    return *up_;
-  }
-
-  policy& active_policy() {
-    return *policy_;
-  }
-
-  size_t max_consecutive_reads() const noexcept {
-    return max_consecutive_reads_;
-  }
-
-  void max_consecutive_reads(size_t value) noexcept {
-    max_consecutive_reads_ = value;
-  }
-
-  // -- implementation of socket_event_layer -----------------------------------
-
-  error start(socket_manager* owner) override;
-
-  socket handle() const override;
-
-  void handle_read_event() override;
-
-  void handle_write_event() override;
-
-  void abort(const error& reason) override;
-
-  bool finalized() const noexcept override;
-
-protected:
-  // -- utility functions ------------------------------------------------------
-
-  /// Consumes as much data from the buffer as possible.
-  void handle_buffered_data();
-
-  /// Calls abort on the upper layer and deregisters the transport from events.
-  void fail(const error& reason);
-
-  // -- member variables -------------------------------------------------------
-
-  /// Stores temporary flags.
-  flags_t flags_;
-
-  /// Caches the config parameter for limiting max. socket operations.
-  size_t max_consecutive_reads_ = defaults::middleman::max_consecutive_reads;
-
-  /// Caches the write buffer size of the socket.
-  uint32_t max_write_buf_size_ = 0;
-
-  /// Stores what the user has configured as read threshold.
-  uint32_t min_read_size_ = 0;
-
-  /// Stores what the user has configured as max. number of bytes to receive.
-  uint32_t max_read_size_ = 0;
-
-  /// Stores how many bytes are currently buffered, i.e., how many bytes from
-  /// `read_buf_` are filled with actual data.
-  size_t buffered_ = 0;
+  virtual byte_buffer& read_buffer() noexcept = 0;
```

Comment:
I couldn't find references to this member function. Do we really need `read_buffer()` and `write_buffer()`?

Reply:
Good catch.